File: parser.js
// parser.js
const axios = require('axios');
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class RateParser {
    constructor(cacheTTL = 60) {
        this.apiURL = 'https://api.exchangerate.host/latest?base=USD';
        this.rates = {};
        this.lastUpdate = null;
        this.cacheTTL = cacheTTL;
        this.history = [];
        this.fetchRates();
    }

    async fetchRates() {
        try {
            const response = await axios.get(this.apiURL, { timeout: 5000 });
            if (response.data.success) {
                this.rates = response.data.rates;
                this.lastUpdate = new Date();
                this.history.push({
                    timestamp: this.lastUpdate.toISOString(),
                    rates: { ...this.rates }
                });
                if (this.history.length > 10) this.history.shift();
                return true;
            }
            return false;
        } catch (error) {
            console.error('Error fetching rates:', error.message);
            return false;
        }
    }

    getRates() {
        if (this.lastUpdate && (Date.now() - this.lastUpdate.getTime()) > this.cacheTTL * 1000) {
            this.fetchRates();
        }
        return this.rates;
    }

    filterRates(currencies) {
        const rates = this.getRates();
        const result = {};
        currencies.forEach(c => {
            if (rates[c] !== undefined) result[c] = rates[c];
        });
        return result;
    }

    searchCurrency(query) {
        const rates = this.getRates();
        const result = {};
        const q = query.toLowerCase();
        Object.keys(rates).forEach(k => {
            if (k.toLowerCase().includes(q)) result[k] = rates[k];
        });
        return result;
    }

    async refresh() {
        return await this.fetchRates();
    }

    exportJSON(filename = 'rates.json') {
        try {
            const data = {
                timestamp: this.lastUpdate ? this.lastUpdate.toISOString() : null,
                rates: this.getRates()
            };
            fs.writeFileSync(filename, JSON.stringify(data, null, 2));
            return true;
        } catch (error) {
            console.error('Export failed:', error.message);
            return false;
        }
    }

    exportCSV(filename = 'rates.csv') {
        try {
            const rates = this.getRates();
            const lines = ['Currency,Rate'];
            Object.keys(rates).sort().forEach(k => {
                lines.push(`${k},${rates[k].toFixed(4)}`);
            });
            fs.writeFileSync(filename, lines.join('\n'));
            return true;
        } catch (error) {
            console.error('Export failed:', error.message);
            return false;
        }
    }

    showHistory() {
        if (this.history.length === 0) {
            console.log('No history yet.');
            return;
        }
        this.history.forEach((entry, i) => {
            console.log(`[${i+1}] ${entry.timestamp} – ${Object.keys(entry.rates).length} currencies`);
        });
    }

    displayRates(rates) {
        if (!rates) rates = this.getRates();
        if (Object.keys(rates).length === 0) {
            console.log('No rates available.');
            return;
        }
        const keys = Object.keys(rates).sort();
        console.log(`\nRates (USD base) – updated: ${this.lastUpdate ? this.lastUpdate.toLocaleString() : 'unknown'}`);
        console.log('-'.repeat(40));
        keys.slice(0, 20).forEach(k => {
            console.log(`${k.padEnd(5)} : ${rates[k].toFixed(4)}`);
        });
        if (keys.length > 20) {
            console.log(`... and ${keys.length - 20} more`);
        }
    }
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function main() {
    const parser = new RateParser();
    console.log('=== Exchange Rate Parser ===');
    while (true) {
        console.log('\n1. Show all rates');
        console.log('2. Filter by currency');
        console.log('3. Search currency');
        console.log('4. Export to JSON');
        console.log('5. Export to CSV');
        console.log('6. Refresh rates');
        console.log('7. Show history');
        console.log('8. Exit');
        const choice = await ask('Choose: ');
        switch (choice.trim()) {
            case '1':
                parser.displayRates();
                break;
            case '2': {
                const input = await ask('Enter currency codes (comma-separated): ');
                const currencies = input.toUpperCase().split(',').map(c => c.trim()).filter(c => c);
                if (currencies.length) {
                    const filtered = parser.filterRates(currencies);
                    parser.displayRates(filtered);
                } else {
                    console.log('No currencies specified.');
                }
                break;
            }
            case '3': {
                const query = await ask('Enter currency code or name: ');
                if (query.trim()) {
                    const found = parser.searchCurrency(query.trim());
                    parser.displayRates(found);
                } else {
                    console.log('Query cannot be empty.');
                }
                break;
            }
            case '4': {
                let fname = await ask('Filename (default: rates.json): ');
                fname = fname.trim() || 'rates.json';
                if (parser.exportJSON(fname)) {
                    console.log(`Exported to ${fname}`);
                }
                break;
            }
            case '5': {
                let fname = await ask('Filename (default: rates.csv): ');
                fname = fname.trim() || 'rates.csv';
                if (parser.exportCSV(fname)) {
                    console.log(`Exported to ${fname}`);
                }
                break;
            }
            case '6':
                if (await parser.refresh()) {
                    console.log('Rates refreshed.');
                } else {
                    console.log('Refresh failed.');
                }
                break;
            case '7':
                parser.showHistory();
                break;
            case '8':
                console.log('Goodbye!');
                rl.close();
                return;
            default:
                console.log('Invalid choice.');
        }
    }
}

main().catch(console.error);

File: Parser.java
// Parser.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.*;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;
import com.google.gson.*;

public class Parser {
    private static final String API_URL = "https://api.exchangerate.host/latest?base=USD";
    private static final int CACHE_TTL = 60;
    private Map<String, Double> rates;
    private Instant lastUpdate;
    private List<HistoryEntry> history;
    private final Gson gson;

    private static class HistoryEntry {
        String timestamp;
        Map<String, Double> rates;
    }

    private static class RateResponse {
        boolean success;
        Map<String, Double> rates;
    }

    public Parser() {
        this.rates = new HashMap<>();
        this.history = new ArrayList<>();
        this.gson = new GsonBuilder().create();
        fetchRates();
    }

    private boolean fetchRates() {
        try {
            URL url = new URL(API_URL);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setConnectTimeout(5000);
            conn.setReadTimeout(5000);

            if (conn.getResponseCode() != 200) {
                System.out.println("HTTP error: " + conn.getResponseCode());
                return false;
            }

            try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                String json = br.lines().collect(Collectors.joining());
                RateResponse response = gson.fromJson(json, RateResponse.class);
                if (response.success && response.rates != null) {
                    rates = response.rates;
                    lastUpdate = Instant.now();
                    HistoryEntry entry = new HistoryEntry();
                    entry.timestamp = lastUpdate.toString();
                    entry.rates = new HashMap<>(rates);
                    history.add(entry);
                    if (history.size() > 10) history.remove(0);
                    return true;
                }
            }
            return false;
        } catch (Exception e) {
            System.out.println("Error fetching rates: " + e.getMessage());
            return false;
        }
    }

    public Map<String, Double> getRates() {
        if (lastUpdate != null && Instant.now().getEpochSecond() - lastUpdate.getEpochSecond() > CACHE_TTL) {
            fetchRates();
        }
        return rates;
    }

    public Map<String, Double> filterRates(List<String> currencies) {
        Map<String, Double> allRates = getRates();
        Map<String, Double> result = new HashMap<>();
        for (String c : currencies) {
            if (allRates.containsKey(c)) result.put(c, allRates.get(c));
        }
        return result;
    }

    public Map<String, Double> searchCurrency(String query) {
        Map<String, Double> allRates = getRates();
        Map<String, Double> result = new HashMap<>();
        String q = query.toLowerCase();
        for (Map.Entry<String, Double> entry : allRates.entrySet()) {
            if (entry.getKey().toLowerCase().contains(q)) {
                result.put(entry.getKey(), entry.getValue());
            }
        }
        return result;
    }

    public boolean refresh() {
        return fetchRates();
    }

    public boolean exportJSON(String filename) {
        try {
            Map<String, Object> data = new HashMap<>();
            data.put("timestamp", lastUpdate != null ? lastUpdate.toString() : null);
            data.put("rates", getRates());
            String json = gson.toJson(data);
            Files.write(Paths.get(filename), json.getBytes());
            return true;
        } catch (Exception e) {
            System.out.println("Export failed: " + e.getMessage());
            return false;
        }
    }

    public boolean exportCSV(String filename) {
        try {
            Map<String, Double> rates = getRates();
            List<String> lines = new ArrayList<>();
            lines.add("Currency,Rate");
            rates.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(e -> lines.add(e.getKey() + "," + String.format("%.4f", e.getValue())));
            Files.write(Paths.get(filename), lines);
            return true;
        } catch (Exception e) {
            System.out.println("Export failed: " + e.getMessage());
            return false;
        }
    }

    public void showHistory() {
        if (history.isEmpty()) {
            System.out.println("No history yet.");
            return;
        }
        for (int i = 0; i < history.size(); i++) {
            System.out.printf("[%d] %s – %d currencies%n", i+1, history.get(i).timestamp, history.get(i).rates.size());
        }
    }

    public void displayRates(Map<String, Double> rates) {
        if (rates == null) rates = getRates();
        if (rates.isEmpty()) {
            System.out.println("No rates available.");
            return;
        }
        System.out.printf("%nRates (USD base) – updated: %s%n", 
            lastUpdate != null ? lastUpdate.toString() : "unknown");
        System.out.println(new String(new char[40]).replace('\0', '-'));
        int count = 0;
        for (Map.Entry<String, Double> entry : rates.entrySet().stream()
                .sorted(Map.Entry.comparingByKey()).collect(Collectors.toList())) {
            if (count >= 20) break;
            System.out.printf("%-5s : %.4f%n", entry.getKey(), entry.getValue());
            count++;
        }
        if (rates.size() > 20) {
            System.out.printf("... and %d more%n", rates.size() - 20);
        }
    }

    public static void main(String[] args) throws IOException {
        Parser parser = new Parser();
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("=== Exchange Rate Parser ===");
        while (true) {
            System.out.println("\n1. Show all rates");
            System.out.println("2. Filter by currency");
            System.out.println("3. Search currency");
            System.out.println("4. Export to JSON");
            System.out.println("5. Export to CSV");
            System.out.println("6. Refresh rates");
            System.out.println("7. Show history");
            System.out.println("8. Exit");
            System.out.print("Choose: ");
            String choice = reader.readLine().trim();
            switch (choice) {
                case "1":
                    parser.displayRates(null);
                    break;
                case "2":
                    System.out.print("Enter currency codes (comma-separated): ");
                    String input = reader.readLine().trim().toUpperCase();
                    List<String> currencies = Arrays.stream(input.split(","))
                        .map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
                    if (!currencies.isEmpty()) {
                        Map<String, Double> filtered = parser.filterRates(currencies);
                        parser.displayRates(filtered);
                    } else {
                        System.out.println("No currencies specified.");
                    }
                    break;
                case "3":
                    System.out.print("Enter currency code or name: ");
                    String query = reader.readLine().trim();
                    if (!query.isEmpty()) {
                        Map<String, Double> found = parser.searchCurrency(query);
                        parser.displayRates(found);
                    } else {
                        System.out.println("Query cannot be empty.");
                    }
                    break;
                case "4":
                    System.out.print("Filename (default: rates.json): ");
                    String fname = reader.readLine().trim();
                    if (fname.isEmpty()) fname = "rates.json";
                    if (parser.exportJSON(fname)) {
                        System.out.println("Exported to " + fname);
                    }
                    break;
                case "5":
                    System.out.print("Filename (default: rates.csv): ");
                    fname = reader.readLine().trim();
                    if (fname.isEmpty()) fname = "rates.csv";
                    if (parser.exportCSV(fname)) {
                        System.out.println("Exported to " + fname);
                    }
                    break;
                case "6":
                    if (parser.refresh()) {
                        System.out.println("Rates refreshed.");
                    } else {
                        System.out.println("Refresh failed.");
                    }
                    break;
                case "7":
                    parser.showHistory();
                    break;
                case "8":
                    System.out.println("Goodbye!");
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}
